### DISCUSSION TITLE
Execution order of fwupdx64.efi and shim as bootx64.efi together with grubx64.efi

### DISCUSSION THREAD

**Discussion by Dvergatal** [fwupd/fwupd#7588] [General]
Hi all,
I have a question regarding the boot flow of of efis with the usage of `fwupd`. Mainly I am using `shim` along with `seloader` to but grub and then kernel. A complete boot flow looks like as following:

    UEFI firmware boot manager (UEFI Secure Boot enabled) ->
        shim (verified by a DB certificate) ->
            SELoader (ditto) ->
                grub (ditto) ->
                    grub.cfg (ditto)
                    kernel (ditto)
                    initramfs (ditto)

According to the documentation of shim it has this `DEFAULT_LOADER` which sets the next efi to load. By default it is set to grub but in my case it is to this seloader efi.

What I'd like to know is, if this may somehow interfere the boot flow when it comes to `fwupdx64.efi`? I was looking for some flow diagrams on how `fwupd` works but I wasn't able to. Can someone can point me on the right direction were to look for the informations or at least explain that to me?

**Comment by hughsie** [comment]
Check out the `efibootmgr` on standard setup, e.g.

    Boot0000* Linux-Firmware-Updater	HD(1,GPT,c7e198db-ec30-4a23-b0fd-8e2dbc092a01,0x800,0x12c000)/\EFI\fedora\shimx64.efi5c00660077007500700064007800360034002e006500660069000000

So here you can see we're using shim to "_chainload_" `fwupdx64.efi` (UCS-4 path!). Maybe https://github.com/fwupd/fwupd/blob/main/plugins/uefi-capsule/README.md#update-behavior helps too?

**Comment by Dvergatal** [comment]
Hi @hughsie,
thx for quick respond. So according to this `fwupdx64.efi` is being run first instead of `bootx64.efi`? Another question which I have is how does `fwupd` daemon relate to `fwupdx64.efi` ? Does the daemon needs the efi to make an update?

**Reply by Dvergatal** [reply]
OK I think I misunderstood it at first. So according to the link given by you, `fwupd` daemon is handling the boot flow in UEFI and changes the order between `fwupdx64.efi` and `shimx64.efi`. Am I understanding it correctly?

**Reply by hughsie** [reply]
Yes, fwupd creates a new BootXXXX key and changes BootNext for nvram updates. For CapsuleOnDisk UEFI updates we don't need to change anything, we can just write a file into the ESP.

**Reply by Dvergatal** [reply]
So the fwupd does (creates new BootXXXX key  and changes BootNext for nvram updates ) that automatically when file is written into the ESP?

One more thing about `shimx64.efi`, because in my case it is named as `bootx64.efi` does it change anything? Meaning if `shimx64.efi` name is somewhere hardcoded?

Regarding capsule for an update it needs to be prepared. Currently I have just a `bios.bin` update blob and `bios.efi` efi updater given by UEFI vendor. In order to run an update I have to run `bios.efi bios.bin` from within UEFI Shell and I have a question is there some manual for preparing such capsule? Also with what keys such a capsule is signed with? Last if I need support from vendor to create this capsule (maybe you have some specs)?

**Reply by hughsie** [reply]
> So the fwupd does that automatically

Yup.

> One more thing about shimx64.efi, because in my case it is named as bootx64.efi does it change anything

I think various places assume shim is called `shim*.efi` in fwupd I'm afraid.

> Currently I have just a bios.bin update blob and bios.efi efi updater given by UEFI vendor

That's not really a capsule, that's not going to work. You need to ask your vendor for a signed capsule rather a raw bios dump.

**Reply by Dvergatal** [reply]
> > So the fwupd does that automatically
> 
> Yup.

Great :D

> > One more thing about shimx64.efi, because in my case it is named as bootx64.efi does it change anything
> 
> I think various places assume shim is called `shim*.efi` in fwupd I'm afraid.

OK so I need to get back to the proper naming...

> > Currently I have just a bios.bin update blob and bios.efi efi updater given by UEFI vendor
> 
> That's not really a capsule, that's not going to work. You need to ask your vendor for a signed capsule rather a raw bios dump.

I know that it's not, which is why I am asking you :P. Mainly because I'm using my own secure boot keys and was thinking how much effort I would need to put into creating such capsule.

Generally speaking, I work in an edge computing project and I need to know what to tell to our UEFI vendor. In addition as I already mentioned above I'm using own secure boot keys and dunno how this capsule is signed, meaning is it only signed with UEFI DBX key or some additional? This is the most important thing for me.

**Reply by hughsie** [reply]
> and I need to know what to tell to our UEFI vendor

Please generate a UEFI capsule signed with the machine PK signing key. If its AMI, Phoenix or Insyde it's just a different checkbox on the UI -- and the capsule is probably already being generated already. You might need to ask forcefully :)

**Reply by Dvergatal** [reply]
> > and I need to know what to tell to our UEFI vendor
> 
> Please generate a UEFI capsule signed with the machine PK signing key.

For signing the UEFI capsule PK key is being used? Not DBX key? Somehow on my intuition it makes sense as UEFI is being updated...

> If its AMI, Phoenix or Insyde it's just a different checkbox on the UI -- and the capsule is probably already being generated already. You might need to ask forcefully :)

We are using Insyde, but if you are saying that this is just one checkbox for capsule generation, then we should not encounter troubles with that hopefully.

BIG thanks for your support. I'm really glad that you have enlightened me :+1:

**Reply by superm1** [reply]
> 
> We are using Insyde, but if you are saying that this is just one checkbox for capsule generation, then we should not encounter troubles with that hopefully.
> 

If you have an `isflash.bin` you should be able to pass this in as a capsule payload to `fwupdtool install-blob`. You can generate the matching XML and put it in to a cab to post to LVFS and use with `fwupdmgr`.

If not; this is _usually_ what Insyde generates when a capsule is made.

**Reply by Dvergatal** [reply]
> > We are using Insyde, but if you are saying that this is just one checkbox for capsule generation, then we should not encounter troubles with that hopefully.
> 
> If you have an `isflash.bin` you should be able to pass this in as a capsule payload to `fwupdtool install-blob`. You can generate the matching XML and put it in to a cab to post to LVFS and use with `fwupdmgr`.

What is `isflash.bin`? Is it normal UEFI update bin blob?

> If not; this is _usually_ what Insyde generates when a capsule is made.

I still need to discuss that with our vendor.

**Reply by superm1** [reply]
> What is isflash.bin? Is it normal UEFI update bin blob?

It's an artifact from the firmware build process Insyde uses.

**Reply by Dvergatal** [reply]
> > What is isflash.bin? Is it normal UEFI update bin blob?
> 
> It's an artifact from the firmware build process Insyde uses.

Yes, but what I was rather asking, how does it differ from normal bios update (meaning the one I'm already using)? I'm asking because maybe it is already the one which I'm using for an update?

**Reply by superm1** [reply]
> > > What is isflash.bin? Is it normal UEFI update bin blob?
> > 
> > It's an artifact from the firmware build process Insyde uses.
> 
> Yes, but what I was rather asking, how does it differ from normal bios update (meaning the one I'm already using)? I'm asking because maybe it is already the one which I'm using for an update?

It has been assembled and signed in a format the capsule updater built on can accept.

**Reply by Dvergatal** [reply]
> > So the fwupd does that automatically
> 
> Yup.
> 
> > One more thing about shimx64.efi, because in my case it is named as bootx64.efi does it change anything
> 
> I think various places assume shim is called `shim*.efi` in fwupd I'm afraid.

I have just build fwupd on my yocto project and I have noticed in `/etc/fwupd/uefi_capsule.conf ` file this paragraph:
```bash
# with the UEFI removable path enabled, the default esp path is set to /EFI/boot
# the shim EFI binary and presumably this is $ESP/EFI/boot/bootx64.efi
#FallbacktoRemovablePath=false
```
and previously you were saying that shim in various places is called `shim*.efi`, but this section denies your quote. How is that then?

Additionally, what I have noticed, is that - I dunno if it is a bug, but from my point of view it is - even if `MACHINE_FEATURES` contains `efi`, `fwupd-efi` package is not installed to rootfs. I need to add it separately into DEPENDS...

**Reply by superm1** [reply]
> > > So the fwupd does that automatically
> > 
> > Yup.
> > 
> > > One more thing about shimx64.efi, because in my case it is named as bootx64.efi does it change anything
> > 
> > I think various places assume shim is called `shim*.efi` in fwupd I'm afraid.
> 
> I have just build fwupd on my yocto project and I have noticed in `/etc/fwupd/uefi_capsule.conf ` file this paragraph:
> ```bash
> # with the UEFI removable path enabled, the default esp path is set to /EFI/boot
> # the shim EFI binary and presumably this is $ESP/EFI/boot/bootx64.efi
> #FallbacktoRemovablePath=false
> ```

This sounds like an old fwupd version. We don't offer options like that in the current configuration file.

https://fwupd.github.io/libfwupdplugin/fwupd.conf.html#uefi_capsule-parameters

> and previously you were saying that shim in various places is called `shim*.efi`, but this section denies your quote. How is that then?
> 

We used to have a way to fall back to the uefi removable path to work around BIOS bugs but it is too fragile and updates don't work consistently across vendors when using that path.

> Additionally, what I have noticed, is that - I dunno if it is a bug, but from my point of view it is - even if `MACHINE_FEATURES` contains `efi`, `fwupd-efi` package is not installed to rootfs. I need to add it separately into DEPENDS... 
> 
> 

This sounds like yocto stuff not fwupd stuff. Talk to yocto people about it.

**Reply by Dvergatal** [reply]
> > > > So the fwupd does that automatically
> > > 
> > > 
> > > Yup.
> > > > One more thing about shimx64.efi, because in my case it is named as bootx64.efi does it change anything
> > > 
> > > 
> > > I think various places assume shim is called `shim*.efi` in fwupd I'm afraid.
> > 
> > 
> > I have just build fwupd on my yocto project and I have noticed in `/etc/fwupd/uefi_capsule.conf ` file this paragraph:
> > ```shell
> > # with the UEFI removable path enabled, the default esp path is set to /EFI/boot
> > # the shim EFI binary and presumably this is $ESP/EFI/boot/bootx64.efi
> > #FallbacktoRemovablePath=false
> > ```
> 
> This sounds like an old fwupd version. We don't offer options like that in the current configuration file.

I'm on Kirkstone LTS release and fwupd is 1.7.6.

> https://fwupd.github.io/libfwupdplugin/fwupd.conf.html#uefi_capsule-parameters
> 
> > and previously you were saying that shim in various places is called `shim*.efi`, but this section denies your quote. How is that then?
> 
> We used to have a way to fall back to the uefi removable path to work around BIOS bugs but it is too fragile and updates don't work consistently across vendors when using that path.

If I understand correctly, this `uefi removable path` was equal `$ESP/EFI/boot/bootx64.efi`?

> > Additionally, what I have noticed, is that - I dunno if it is a bug, but from my point of view it is - even if `MACHINE_FEATURES` contains `efi`, `fwupd-efi` package is not installed to rootfs. I need to add it separately into DEPENDS...
> 
> This sounds like yocto stuff not fwupd stuff. Talk to yocto people about it.

Yes it is:P I though that maybe you were responsible for creation of the recipe. Sorry for that.

**Reply by superm1** [reply]
> > > > > So the fwupd does that automatically
> > > > 
> > > > 
> > > > Yup.
> > > > > One more thing about shimx64.efi, because in my case it is named as bootx64.efi does it change anything
> > > > 
> > > > 
> > > > I think various places assume shim is called `shim*.efi` in fwupd I'm afraid.
> > > 
> > > 
> > > I have just build fwupd on my yocto project and I have noticed in `/etc/fwupd/uefi_capsule.conf ` file this paragraph:
> > > ```shell
> > > # with the UEFI removable path enabled, the default esp path is set to /EFI/boot
> > > # the shim EFI binary and presumably this is $ESP/EFI/boot/bootx64.efi
> > > #FallbacktoRemovablePath=false
> > > ```
> > 
> > This sounds like an old fwupd version. We don't offer options like that in the current configuration file.
> 
> I'm on Kirkstone LTS release and fwupd is 1.7.6.

It's weird to me for an "LTS" to use EOL software. You should upgrade.

https://github.com/fwupd/fwupd/blob/main/SECURITY.md

> 
> > https://fwupd.github.io/libfwupdplugin/fwupd.conf.html#uefi_capsule-parameters
> > 
> > > and previously you were saying that shim in various places is called `shim*.efi`, but this section denies your quote. How is that then?
> > 
> > We used to have a way to fall back to the uefi removable path to work around BIOS bugs but it is too fragile and updates don't work consistently across vendors when using that path.
> 
> If I understand correctly, this `uefi removable path` was equal `$ESP/EFI/boot/bootx64.efi`?

Yes. It is a house of cards, only **intended** to be used with removable media.

Using it on your HDD is going to have different results for different vendors.

> 
> > > Additionally, what I have noticed, is that - I dunno if it is a bug, but from my point of view it is - even if `MACHINE_FEATURES` contains `efi`, `fwupd-efi` package is not installed to rootfs. I need to add it separately into DEPENDS...
> > 
> > This sounds like yocto stuff not fwupd stuff. Talk to yocto people about it.
> 
> Yes it is:P I though that maybe you were responsible for creation of the recipe. Sorry for that.
> 
>

**Reply by Dvergatal** [reply]
> > > > > > So the fwupd does that automatically
> > > > > 
> > > > > 
> > > > > Yup.
> > > > > > One more thing about shimx64.efi, because in my case it is named as bootx64.efi does it change anything
> > > > > 
> > > > > 
> > > > > I think various places assume shim is called `shim*.efi` in fwupd I'm afraid.
> > > > 
> > > > 
> > > > I have just build fwupd on my yocto project and I have noticed in `/etc/fwupd/uefi_capsule.conf ` file this paragraph:
> > > > ```shell
> > > > # with the UEFI removable path enabled, the default esp path is set to /EFI/boot
> > > > # the shim EFI binary and presumably this is $ESP/EFI/boot/bootx64.efi
> > > > #FallbacktoRemovablePath=false
> > > > ```
> > > 
> > > 
> > > This sounds like an old fwupd version. We don't offer options like that in the current configuration file.
> > 
> > 
> > I'm on Kirkstone LTS release and fwupd is 1.7.6.
> 
> It's weird to me for an "LTS" to use EOL software. You should upgrade.
> 
> https://github.com/fwupd/fwupd/blob/main/SECURITY.md

Thx for that. I wasn't really aware of it...

> 
> > > https://fwupd.github.io/libfwupdplugin/fwupd.conf.html#uefi_capsule-parameters
> > > > and previously you were saying that shim in various places is called `shim*.efi`, but this section denies your quote. How is that then?
> > > 
> > > 
> > > We used to have a way to fall back to the uefi removable path to work around BIOS bugs but it is too fragile and updates don't work consistently across vendors when using that path.
> > 
> > 
> > If I understand correctly, this `uefi removable path` was equal `$ESP/EFI/boot/bootx64.efi`?
> 
> Yes. It is a house of cards, only **intended** to be used with removable media.
> 
> Using it on your HDD is going to have different results for different vendors.
> 
> > > > Additionally, what I have noticed, is that - I dunno if it is a bug, but from my point of view it is - even if `MACHINE_FEATURES` contains `efi`, `fwupd-efi` package is not installed to rootfs. I need to add it separately into DEPENDS...
> > > 
> > > 
> > > This sounds like yocto stuff not fwupd stuff. Talk to yocto people about it.
> > 
> > 
> > Yes it is:P I though that maybe you were responsible for creation of the recipe. Sorry for that.

**Reply by Dvergatal** [reply]
I don't want to quote so many sentences:P and would like to ask few more questions.

I see that `fwupd.service` is launched through systemd via D-Bus automatically whenever it is needed and I've noticed that the service is not enabled permanently but rather launched in example on my Gentoo machine with KDE by a discovery app. But how is it behaving with this capsule update on-disk and `\EFI\UpdateCapsule` directory? Because if the service is turned off and daemon is not running at all, how will it discover that there was put a capsule file to that directory?

Another thing is question why `fwupdx64.efi` is installed in this location `/usr/libexec/fwupd/efi/fwupdx64.efi` and not in `$ESP`?Because I have noticed in some old threads that it was previously.

**Reply by superm1** [reply]
> I see that fwupd.service is launched through systemd via D-Bus automatically whenever it is needed and I've noticed that the service is not enabled permanently but rather launched in example on my Gentoo machine with KDE by a discovery app. 

Right.  It will quit after being idle for some (configurable) time as well.

> But how is it behaving with this capsule update on-disk and \EFI\UpdateCapsule directory? Because if the service is turned off and daemon is not running at all, how will it discover that there was put a capsule file to that directory?

The payload is staged on the ESP, and will be applied on the next reboot.  When using CoD the firmware knows to pick it up (from `OsIndications`).  When using NVRAM the boot variable pointing to the firmware updater EFI application is launched on next boot.

> Another thing is question why fwupdx64.efi is installed in this location /usr/libexec/fwupd/efi/fwupdx64.efi and not in $ESP?Because I have noticed in some old threads that it was previously.

The fwupd EFI application can be updated independently of the fwupd linux application.  Because of this the daemon will check whats on the ESP and update it if necessary.  You can also update it yourself manually by packaging or your OS build.

**Reply by Dvergatal** [reply]
> > I see that fwupd.service is launched through systemd via D-Bus automatically whenever it is needed and I've noticed that the service is not enabled permanently but rather launched in example on my Gentoo machine with KDE by a discovery app.
> 
> Right. It will quit after being idle for some (configurable) time as well.
> 
> > But how is it behaving with this capsule update on-disk and \EFI\UpdateCapsule directory? Because if the service is turned off and daemon is not running at all, how will it discover that there was put a capsule file to that directory?
> 
> The payload is staged on the ESP, and will be applied on the next reboot. When using CoD the firmware knows to pick it up (from `OsIndications`). When using NVRAM the boot variable pointing to the firmware updater EFI application is launched on next boot.

Ok I will explain. My payload will be not delivered from LVFS, but swupdate firmware update image will put capsule into ESP and how does the daemon will get to now that it has an update?

> 
> > Another thing is question why fwupdx64.efi is installed in this location /usr/libexec/fwupd/efi/fwupdx64.efi and not in $ESP?Because I have noticed in some old threads that it was previously.
> 
> The fwupd EFI application can be updated independently of the fwupd linux application. Because of this the daemon will check whats on the ESP and update it if necessary. You can also update it yourself manually by packaging or your OS build.

Yes but fwupd EFI is being installed into `/usr/libexec/fwupd/efi/fwupdx64.efi` not $ESP and that concerns me, why?

**Reply by superm1** [reply]
> > > I see that fwupd.service is launched through systemd via D-Bus automatically whenever it is needed and I've noticed that the service is not enabled permanently but rather launched in example on my Gentoo machine with KDE by a discovery app.
> > 
> > Right. It will quit after being idle for some (configurable) time as well.
> > 
> > > But how is it behaving with this capsule update on-disk and \EFI\UpdateCapsule directory? Because if the service is turned off and daemon is not running at all, how will it discover that there was put a capsule file to that directory?
> > 
> > The payload is staged on the ESP, and will be applied on the next reboot. When using CoD the firmware knows to pick it up (from `OsIndications`). When using NVRAM the boot variable pointing to the firmware updater EFI application is launched on next boot.
> 
> Ok I will explain. My payload will be not delivered from LVFS, but swupdate firmware update image will put capsule into ESP and how does the daemon will get to now that it has an update?

The way you're doing it doesn't give the user any control.

If you are delivering it via a local file it should be in a local fwupd remote not directly on the ESP.

Then the user can choose to install it from their client which would copy it to the ESP at that time.

> 
> > 
> > > Another thing is question why fwupdx64.efi is installed in this location /usr/libexec/fwupd/efi/fwupdx64.efi and not in $ESP?Because I have noticed in some old threads that it was previously.
> > 
> > The fwupd EFI application can be updated independently of the fwupd linux application. Because of this the daemon will check whats on the ESP and update it if necessary. You can also update it yourself manually by packaging or your OS build.
> 
> Yes but fwupd EFI is being installed into `/usr/libexec/fwupd/efi/fwupdx64.efi` not $ESP and that concerns me, why?
> 
> 

It's due to the lifecycle of this EFI application.
Say for example that you needed to do a key rotation with the key you signed the binary with. Most packaging systems can't directly install across filesystems but only to the rootfs.

It's put there by the packaging by default. You can copy it to the ESP manually if you don't want the daemon to do it.

**Reply by Dvergatal** [reply]
> > > > I see that fwupd.service is launched through systemd via D-Bus automatically whenever it is needed and I've noticed that the service is not enabled permanently but rather launched in example on my Gentoo machine with KDE by a discovery app.
> > > 
> > > 
> > > Right. It will quit after being idle for some (configurable) time as well.
> > > > But how is it behaving with this capsule update on-disk and \EFI\UpdateCapsule directory? Because if the service is turned off and daemon is not running at all, how will it discover that there was put a capsule file to that directory?
> > > 
> > > 
> > > The payload is staged on the ESP, and will be applied on the next reboot. When using CoD the firmware knows to pick it up (from `OsIndications`). When using NVRAM the boot variable pointing to the firmware updater EFI application is launched on next boot.
> > 
> > 
> > Ok I will explain. My payload will be not delivered from LVFS, but swupdate firmware update image will put capsule into ESP and how does the daemon will get to now that it has an update?
> 
> The way you're doing it doesn't give the user any control.
> 
> If you are delivering it via a local file it should be in a local fwupd remote not directly on the ESP.
> 
> Then the user can choose to install it from their client which would copy it to the ESP at that time.

For example firmware to `/usr/share/fwupd/remotes.d/vendor/firmware` and metadata to `/usr/share/fwupd/remotes.d/vendor/vendor.xml.gz`?

> 
> > > > Another thing is question why fwupdx64.efi is installed in this location /usr/libexec/fwupd/efi/fwupdx64.efi and not in $ESP?Because I have noticed in some old threads that it was previously.
> > > 
> > > 
> > > The fwupd EFI application can be updated independently of the fwupd linux application. Because of this the daemon will check whats on the ESP and update it if necessary. You can also update it yourself manually by packaging or your OS build.
> > 
> > 
> > Yes but fwupd EFI is being installed into `/usr/libexec/fwupd/efi/fwupdx64.efi` not $ESP and that concerns me, why?
> 
> It's due to the lifecycle of this EFI application. Say for example that you needed to do a key rotation with the key you signed the binary with. Most packaging systems can't directly install across filesystems but only to the rootfs.

Got it :+1: 

> 
> It's put there by the packaging by default. You can copy it to the ESP manually if you don't want the daemon to do it.

I'd like to use daemon, but I still need to make some clear ups, because I don't fully understand everything yet.

**Reply by superm1** [reply]
> > If you are delivering it via a local file it should be in a local fwupd remote not directly on the ESP.
> > 
> > Then the user can choose to install it from their client which would copy it to the ESP at that time.
> 
> For example firmware to `/usr/share/fwupd/remotes.d/vendor/firmware` and metadata to `/usr/share/fwupd/remotes.d/vendor/vendor.xml.gz`?
> 

Yes; here is docs:

https://lvfs.readthedocs.io/en/latest/offline.html#deploy-in-immutable-image


> 
> > 
> > It's put there by the packaging by default. You can copy it to the ESP manually if you don't want the daemon to do it.
> 
> I'd like to use daemon, but I still need to make some clear ups, because I don't fully understand everything yet.
> 
> 

If you're going to use the daemon but not offer the user control I suggest:

1. Place a cab in a local remote as described above
2. you can script `fwupdmgr update --json` as a systemd unit that runs after startup is done (so you don't slow the startup just for this)
3. Next time system reboots it will install firmware.

I will warn though, this experience is pretty jolting to an end user. The next bootup will be a lot longer and if you don't have a good recovery solution and they unplugged the power during the update you can brick the machine.

So I think it's a lot better for you to let user know it's happening.

**Reply by Dvergatal** [reply]
> > > If you are delivering it via a local file it should be in a local fwupd remote not directly on the ESP.
> > > Then the user can choose to install it from their client which would copy it to the ESP at that time.
> > 
> > 
> > For example firmware to `/usr/share/fwupd/remotes.d/vendor/firmware` and metadata to `/usr/share/fwupd/remotes.d/vendor/vendor.xml.gz`?
> 
> Yes; here is docs:
> 
> https://lvfs.readthedocs.io/en/latest/offline.html#deploy-in-immutable-image

This is exactly what I was thinking of but at the very beginning I thought that instead of using LVFS I'm just putting cab file in ESP and daemon does the rest.

> > > It's put there by the packaging by default. You can copy it to the ESP manually if you don't want the daemon to do it.
> > 
> > 
> > I'd like to use daemon, but I still need to make some clear ups, because I don't fully understand everything yet.
> 
> If you're going to use the daemon but not offer the user control I suggest:
> 
>     1. Place a cab in a local remote as described above
> 
>     2. you can script `fwupdmgr update --json` as a systemd unit that runs after startup is done (so you don't slow the startup just for this)

First what is that command for, second instead of creating a disposable systemd unit I was rather thinking of running cab update command from a script which would be inside swu.

P.S. Forgotten to ask, how does `fwupdmgr update` differ from `fwupdtool install-blob` ?

> 
>     3. Next time system reboots it will install firmware.
> 
> 
> I will warn though, this experience is pretty jolting to an end user. The next bootup will be a lot longer and if you don't have a good recovery solution and they unplugged the power during the update you can brick the machine.

What do you mean by 'jolting'? That it may take too long? About bricking the machine, with the current solution, with this efi script it may still happen...

> So I think it's a lot better for you to let user know it's happening.

I know btw. isn't the update showing some sort of a screen with a progress bar?

**Reply by Dvergatal** [reply]
> P.S. Forgotten to ask, how does `fwupdmgr update` differ from `fwupdtool install-blob` ?

Does the difference between both of these is that `fwupdmgr update` is for `*.cab` file and  `fwupdtool install-blob`  for `*.cap` ?

**Reply by hughsie** [reply]
>  fwupdtool install-blob for *.cap

Yes, in general `fwupdtool` is used for debugging and development, and `fwupdmgr` (which talks to the `fwupd` dbus daemon) is used by end users in production.

**Reply by Dvergatal** [reply]
I have to admit that this project is really huge and gives a lot of possibilities. ~~Btw. what are the differences between `*.cab` and '*.cap` files? Where can I read about this?~~

> > We are using Insyde, but if you are saying that this is just one checkbox for capsule generation, then we should not encounter troubles with that hopefully.
> 
> If you have an `isflash.bin` you should be able to pass this in as a capsule payload to `fwupdtool install-blob`. You can generate the matching XML and put it in to a cab to post to LVFS and use with `fwupdmgr`.
> 
> If not; this is _usually_ what Insyde generates when a capsule is made.

I think finally I get this comment :smiley: , but is there a nice tool to create cab? Ah there is https://lvfs.readthedocs.io/en/latest/upload.html

and according to this documentation cab files may be signed with GPG or PKCS#7 and cap files are suppoused to be signed with PK key right?

**Reply by Dvergatal** [reply]
I was trying to test `fwupdtool install-blob WDE5EA1.BIN a45df35ac0e948ee180fe216a5f703f32dda163f`. The device id has been retrieved from `fwupdmgr get-devices` and it has failed. The output of the command looks like below:

```bash
root@eg /home/admins/admin> fwupdtool install-blob WDE5EA1.BIN a45df35ac0e948ee180fe216a5f703f32dda163f
Loading?                 [-                                      ]19:21:53:0140 FuPluginUefiCapsule  SMBIOS BIOS Characteristics Extension Byte 2 is invalid -- UEFI Specification is unsupported, but /sys/firmware/efi exists: System does not support UEFI mode
19:21:59:0546 FuEngine             failed to add device /sys/devices/pci0000:00/0000:00:00.0: failed to open port: Function not implemented
19:21:59:0745 FuEngine             failed to add device /sys/devices/pci0000:00/0000:00:1f.0: failed to open port: Function not implemented
Loading?                 [***************************************]
WARNING: This package has not been validated, it may not work properly.
Decompressing?           [ -                                     ]failed to read EFI variable: Error opening file /sys/firmware/efi/efivars/OsIndications-8be4df61-93ca-11d2-aa0d-00e098032b8c: No such file or directory
```

I ckecked if /sys/firmware/efi/efivars/OsIndications-8be4df61-93ca-11d2-aa0d-00e098032b8c is missing and it is. So my question is if our UEFI implementation is missing support for `fwupd install-blob` ?

**Reply by hughsie** [reply]
> I was trying to test fwupdtool install-blob WDE5EA1.BIN

Is `WDE5EA1.BIN` a capsule?

> I ckecked if /sys/firmware/efi/efivars/OsIndications-8be4df61-93ca-11d2-aa0d-00e098032b8c is missing and it is

I think that's fine; fwupd from git just shows a debug line not a warning -- what fwupd version are you testing with?

**Reply by Dvergatal** [reply]
> > I was trying to test fwupdtool install-blob WDE5EA1.BIN
> 
> Is `WDE5EA1.BIN` a capsule?

In general  `WDE5EA1.BIN` is a flash update which we are using to update from efishell like in below README from vendor:

```bash
Please update Insyde BIOS with following parameters for all intel Apollo Lake series!

For UEFI Mode update :

For example : Shell> fs0:
            : fs0:\>H2OFFT-Sx64.efi 2I390CE1.BIN -all
```

And I thought that maybe according to what @superm1 said:

> If you have an isflash.bin you should be able to pass this in as a capsule payload to fwupdtool install-blob. You can generate the matching XML and put it in to a cab to post to LVFS and use with fwupdmgr.

it might work.

> 
> > I ckecked if /sys/firmware/efi/efivars/OsIndications-8be4df61-93ca-11d2-aa0d-00e098032b8c is missing and it is
> 
> I think that's fine; fwupd from git just shows a debug line not a warning -- what fwupd version are you testing with?

As I have written already, 1.7.6, but I can always update it to the latest one

**Reply by superm1** [reply]
> it might work.

Is it a renamed `isflash.bin` artifact?  If not; it's probably not a capsule.

> As I have written already, 1.7.6, but I can always update it to the latest one

We don't work with EOL software versions upstream, please upgrade.

**Reply by Dvergatal** [reply]
> > it might work.
> 
> Is it a renamed `isflash.bin` artifact? If not; it's probably not a capsule.

I need to talk with our vendor guys.

> > As I have written already, 1.7.6, but I can always update it to the latest one
> 
> We don't work with EOL software versions upstream, please upgrade.

OK I will update first.

I will come back to you after I will get any informations about the capsule and when I will do the update.

**Comment by hughsie** [comment]
> My payload will be not delivered from LVFS, but swupdate firmware update image will put capsule into ESP

It sounds like you need to set up BootXXXX manually -- or install it somewhere else, and then `fwupdtool install-blob` it.

**Reply by superm1** [reply]
> > My payload will be not delivered from LVFS, but swupdate firmware update image will put capsule into ESP
> 
> It sounds like you need to set up BootXXXX manually -- or install it somewhere else, and then `fwupdtool install-blob` it.

I think there is a fundamental question to answer here as yocto was mentioned and this sounds really custom.

Is the plan for the user to be in control of installing the BIOS update or is it going to be invisible?

If it's intended to be invisible then I would question if fwupd really even matters here or if shell script is better to use.

**Reply by Dvergatal** [reply]
> > > My payload will be not delivered from LVFS, but swupdate firmware update image will put capsule into ESP
> > 
> > 
> > It sounds like you need to set up BootXXXX manually -- or install it somewhere else, and then `fwupdtool install-blob` it.
> 
> I think there is a fundamental question to answer here as yocto was mentioned and this sounds really custom.
> 
> Is the plan for the user to be in control of installing the BIOS update or is it going to be invisible?

Generally as I see really huge potential in fwupd features with LVFS and so on I suspect that in the future releases of our OS we will probably want to support them as well, but for the current state I just need to make an update with this approach and coming to your next sentence...

> If it's intended to be invisible then I would question if fwupd really even matters here or if shell script is better to use.

... this is mostly why it can't be set up with this shell script, as I want to have secure boot enabled and shell script will throw me an error about running any efi from it. Because I have already tested this solution and with secure boot it will not work.

**Reply by Dvergatal** [reply]
> > My payload will be not delivered from LVFS, but swupdate firmware update image will put capsule into ESP
> 
> It sounds like you need to set up BootXXXX manually -- or install it somewhere else, and then `fwupdtool install-blob` it.

So `fwupdtool install-blob` does the thing with setting up BootXXXX and installs the blob on $ESP\EFI\UpdateCapsule and after reboot the `fwupdx64.efi` is running?

**Reply by superm1** [reply]
> So fwupdtool install-blob does the thing with setting up BootXXXX and installs the blob on $ESP\EFI\UpdateCapsule and after reboot the fwupdx64.efi is running?

That's correct; both `fwupdtool install-blob` and the `fwupd` daemon would do this action.  Think of `fwupdtool` as the development tool and `fwupd` as the production way to use it.

**Comment by hughsie** [comment]
> I'm just putting cab file in ESP

Not in the ESP. The ESP is where the capsule gets put by fwupd -- the cabinet archive (with the metadata, and the capsule) gets put somewhere on the filesystem (**not** the ESP) so that fwupd can deploy the capsule onto the ESP (along with fwupdx64.efi) when needed.

> Because I have already tested this solution and with secure boot it will not work.

Sure, by design. You can't sign the "flash any ROM blob" for secure boot as it would then allow you to flash any firmware that didn't do secure boot checks. You have to bundle the firmware payload into the updater as a capsule, then sign **that**.

**Reply by Dvergatal** [reply]
> > I'm just putting cab file in ESP
> 
> Not in the ESP. The ESP is where the capsule gets put by fwupd -- the cabinet archive (with the metadata, and the capsule) gets put somewhere on the filesystem (**not** the ESP) so that fwupd can deploy the capsule onto the ESP (along with fwupdx64.efi) when needed.

Yes we have already agreed on that with @superm1 in previous comments above.

> 
> > Because I have already tested this solution and with secure boot it will not work.
> 
> Sure, by design. You can't sign the "flash any ROM blob" for secure boot as it would then allow you to flash any firmware that didn't do secure boot checks. You have to bundle the firmware payload into the updater as a capsule, then sign **that**.

Yes indeed, which is exactly why I need to use fwupd not the ugly approach with the script...

**Comment by Dvergatal** [comment]
One more question to the topic as I dunno if I have asked about it or not and there is already plenty of information in here but when using shim does the fwupd should be signed with DB key or shim key?

**Reply by superm1** [reply]
"Normally" it's signed by the key embedded in shim. If you want to directly sign with a key in db and use without shim that's viable though.
